File: src/java/project2/Tree.java
package project2;

import java.util.Objects;

/**
 * Represents a tree in New York City.
 * 
 * @author Ishan Pranav
 */
public class Tree implements Comparable<Tree>, Species {
    private static final String MANHATTAN_BOROUGH = "Manhattan";

    /**
     * Gets a collection of valid borough names.
     * 
     * @return The valid borough names.
     */
    public static String[] getBoroughs() {
        return new String[] { MANHATTAN_BOROUGH, "Bronx", "Brooklyn", "Queens", "Staten Island" };
    }

    private int treeID;
    private String status;
    private String health;
    private String latinName;
    private String commonName;
    private int zipCode;
    private String borough = MANHATTAN_BOROUGH;
    private double x;
    private double y;

    /**
     * Initializes a new instance of the {@link Tree} class.
     * 
     * @param treeID  the tree identifier.
     * @param species the tree species.
     * @throws IllegalArgumentException species is null.
     */
    public Tree(int treeID, TreeSpecies species) throws IllegalArgumentException {
        setTreeID(treeID);

        if (species == null) {
            throw new IllegalArgumentException("Value cannot be null. Argument name: species.");
        } else {
            setLatinName(species.getLatinName());
            setCommonName(species.getCommonName());
        }
    }

    /**
     * Gets the tree identifier. This property represents the {@code tree_id} field
     * in the dataset.
     * 
     * @return A non-negative integer identifier.
     */
    public int getTreeID() {
        return treeID;
    }

    /**
     * Sets the tree identifier. This property represents the {@code tree_id} field
     * in the dataset.
     * 
     * @param treeID a non-negative integer identifier.
     * @throws IllegalArgumentException if treeID is negative.
     */
    public void setTreeID(int treeID) {
        if (treeID < 0) {
            throw new IllegalArgumentException(
                    "Value is out of range. A non-negative number is required. Argument name: treeID.");
        } else {
            this.treeID = treeID;
        }
    }

    /**
     * Gets the status of the tree. This property represents the {@code status}
     * field in the dataset.
     * 
     * @return {@code "Alive"}, {@code "Dead"}, {@code "Stump"}, {@code null}, or an
     *         empty string to represent the status. Status values are
     *         case-insensitive.
     */
    public String getStatus() {
        return status;
    }

    /**
     * Sets the status of the tree. This property represents the {@code status}
     * field in the dataset.
     * 
     * @param status {@code "Alive"}, {@code "Dead"}, {@code "Stump"}, {@code null},
     *               or an empty string to represent the status. Status values are
     *               case-insensitive.
     * @throws IllegalArgumentException if status is not a valid status.
     */
    public void setStatus(String status) {
        if (status == null || status.equals("") || status.equalsIgnoreCase("Alive") || status.equalsIgnoreCase("Dead")
                || status.equalsIgnoreCase("Stump")) {
            this.status = status;
        } else {
            throw new IllegalArgumentException(
                    "A valid status value is required. Allowed values: null or empty, Alive, Dead, and Stump. Argument name: status.");
        }
    }

    /**
     * Gets the health of the tree. This property represents the {@code health}
     * field in the dataset.
     * 
     * @return {@code "Good"}, {@code "Fair"}, {@code "Poor"}, {@code null}, or an
     *         empty string to represent the health. Health descriptions are
     *         case-insensitive.
     */
    public String getHealth() {
        return health;
    }

    /**
     * Sets the health of the tree. This property represents the {@code health}
     * field in the dataset.
     * 
     * @param health {@code "Good"}, {@code "Fair"}, {@code "Poor"}, {@code null},
     *               or an empty string to represent the health. Health descriptions
     *               are case-insensitive.
     * @throws IllegalArgumentException if health is not a valid health description.
     */
    public void setHealth(String health) {
        if (health == null || health.equals("") || health.equalsIgnoreCase("Good") || health.equalsIgnoreCase("Fair")
                || health.equalsIgnoreCase("Poor")) {
            this.health = health;
        } else {
            throw new IllegalArgumentException(
                    "A valid health value is required. Allowed values: null or empty, Good, Fair, and Poor. Argument name: health.");
        }
    }

    /** {@inheritDoc} */
    public String getLatinName() {
        return latinName;
    }

    /**
     * Sets the scientific (Latin) name of the tree's species. This property
     * represents the {@code spc_latin} field in the dataset.
     * 
     * @param latinName the scientific name, a non-null string.
     * @throws IllegalArgumentException if latinName is null.
     */
    public void setLatinName(String latinName) {
        if (latinName == null) {
            throw new IllegalArgumentException("Value cannot be null. Argument name: latinName.");
        } else {
            this.latinName = latinName;
        }
    }

    /** {@inheritDoc} */
    public String getCommonName() {
        return commonName;
    }

    /**
     * Sets the common (English) name of the tree's species. This property
     * represents the {@code spc_common} field in the dataset.
     * 
     * @param commonName the common name, a non-null string.
     * @throws IllegalArgumentException if commonName is null.
     */
    public void setCommonName(String commonName) {
        if (commonName == null) {
            throw new IllegalArgumentException("Value cannot be null. Argument name: commonName.");
        } else {
            this.commonName = commonName;
        }
    }

    /**
     * Gets the postal (ZIP) code in which the tree is located. This property
     * represents the {@code zipcode} field in the dataset.
     * 
     * @return The postcode, an integer between 00000 and 99999.
     */
    public int getZipCode() {
        return zipCode;
    }

    /**
     * Sets the postal (ZIP) code in which the tree is located. This property
     * represents the {@code zipcode} field in the dataset.
     * 
     * @param zipCode the postcode, an integer between 00000 and 99999.
     * @throws IllegalArgumentException if zipCode is less than 0 or greater than
     *                                  99999.
     */
    public void setZipCode(int zipCode) {
        if (zipCode >= 0 && zipCode <= 99999) {
            this.zipCode = zipCode;
        } else {
            throw new IllegalArgumentException(
                    "Value is out of range. Zip codes must be between 00000 and 99999, inclusive. Argument name: zipCode.");
        }
    }

    /**
     * Gets the borough name in which the tree is located. This property represents
     * the {@code boroname} field in the dataset.
     * 
     * @return {@code "Manhattan"}, {@code "Bronx"}, {@code "Brooklyn"},
     *         {@code "Queens"}, or {@code "Staten Island"} to represent the
     *         borough. Borough names are case-insensitive.
     */
    public String getBorough() {
        return borough;
    }

    /**
     * Sets the borough name in which the tree is located. This property represents
     * the {@code boroname} field in the dataset.
     * 
     * @param borough {@code "Manhattan"}, {@code "Bronx"}, {@code "Brooklyn"},
     *                {@code "Queens"}, or {@code "Staten Island"} to represent the
     *                borough. Borough names are case-insensitive.
     * @throws IllegalArgumentException if borough is not a valid borough name.
     */
    public void setBorough(String borough) {
        for (String item : getBoroughs()) {
            if (item.equalsIgnoreCase(borough)) {
                this.borough = item;

                return;
            }
        }

        throw new IllegalArgumentException(
                "A valid borough name is required. Allowed values: Manhattan, Bronx, Brooklyn, Queens, Staten Island. Argument name: borough.");
    }

    /**
     * Gets the x-coordinate of the tree. This property represents the {@code x_sp}
     * field in the dataset.
     * 
     * @return The x-coordinate.
     */
    public double getX() {
        return x;
    }

    /**
     * Sets the x-coordinate of the tree. This property represents the {@code x_sp}
     * field in the dataset.
     * 
     * @param x the x-coordinate.
     */
    public void setX(double x) {
        this.x = x;
    }

    /**
     * Gets the y-coordinate of the tree. This property represents the {@code y_sp}
     * field in the dataset.
     * 
     * @return the y-coordinate.
     */
    public double getY() {
        return y;
    }

    /**
     * Sets the y-coordinate of the tree. This property represents the {@code y_sp}
     * field in the dataset.
     * 
     * @param y the y-coordinate.
     */
    public void setY(double y) {
        this.y = y;
    }

    /** {@inheritDoc} */
    public int compareTo(Tree o) {
        Tree other = (Tree) o;
        int result = commonName.compareToIgnoreCase(other.commonName);

        if (result == 0) {
            Comparable<Integer> comparable = treeID;

            result = comparable.compareTo(other.treeID);
        }

        return result;
    }

    /** {@inheritDoc} */
    @Override
    public boolean equals(Object obj) {
        if (obj == null || !(obj instanceof Tree)) {
            return false;
        } else {
            Tree other = (Tree) obj;

            return treeID == other.treeID && latinName.equalsIgnoreCase(other.latinName)
                    && commonName.equalsIgnoreCase(other.commonName);
        }
    }

    /** {@inheritDoc} */
    @Override
    public int hashCode() {
        return Objects.hash(treeID, latinName, commonName);
    }

    /** {@inheritDoc} */
    @Override
    public String toString() {
        return String.format("%s (%s) #%d", commonName, latinName, treeID);
    }
}

File: src/cs/TreeCensus/Tree.cs
using System;
using CsvHelper.Configuration.Attributes;

namespace TreeCensus
{
    /// <summary>
    /// Represents a tree.
    /// </summary>
    public partial class Tree : IComparable, IComparable<Tree>, IEquatable<Tree>, IFormattable, ISpecies
    {
        private int _id;

        /// <summary>
        /// Gets or sets the tree identifier.
        /// </summary>
        /// <value>The tree identifier.</value>
        /// <exception cref="System.ArgumentOutOfRangeException">The value is less than 0.</exception>
        [Name("tree_id")]
        public int Id
        {
            get
            {
                return _id;
            }
            set
            {
                if (value < 0)
                {
                    throw new ArgumentOutOfRangeException(ArgumentOutOfRangeException);
                }
                else
                {
                    _id = value;
                }
            }
        }

        /// <summary>
        /// Gets or sets the tree status.
        /// </summary>
        /// <value>The tree status.</value>
        [Name("status")]
        public Status Status { get; set; }

        /// <summary>
        /// Gets or sets the tree health.
        /// </summary>
        /// <value>The tree health.</value>
        [Name("health")]
        public Health Health { get; set; }

        /// <summary>
        /// Gets or sets the scientific (Latin) name of the tree. The default is <see cref="string.Empty"/>.
        /// </summary>
        /// <value>The scientific name.</value>
        [Name("spc_latin")]
        public string LatinName { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the common (English) name of the tree. The default is <see cref="string.Empty"/>.
        /// </summary>
        /// <value>The common name.</value>
        [Name("spc_common")]
        public string CommonName { get; set; } = string.Empty;

        private int _postcode;

        /// <summary>
        /// Gets or sets the postcode.
        /// </summary>
        /// <value>The postcode, a five-digit number between 00000 and 99999.</value>
        /// <exception cref="System.ArgumentOutOfRangeException">The postcode is less than 0 or greater than 99999.</exception>
        [Name("postcode")]
        public int Postcode
        {
            get
            {
                return _postcode;
            }
            set
            {
                if (value >= 0 && value <= 99999)
                {
                    _postcode = value;
                }
                else
                {
                    throw new ArgumentOutOfRangeException(nameof(value), "Postcode must be between 00000 and 99999, inclusive.");
                }
            }
        }

        /// <summary>
        /// Gets or sets the borough in which the tree is located.
        /// </summary>
        /// <value>The borough.</value>
        [Name("borough")]
        public Borough Borough { get; set; }

        /// <summary>
        /// Gets or sets the x-coordinate of the tree.
        /// </summary>
        /// <value>The x-coordinate.</value>
        [Name("x_sp")]
        public double X { get; set; }

        /// <summary>
        /// Gets or sets the y-coordinate of the tree.
        /// </summary>
        /// <value>The y-coordinate.</value>
        [Name("y_sp")]
        public double Y { get; set; }

        /// <summary>
        /// Initializes a new instance of the <see cref="Tree"/> class.
        /// </summary>
        public Tree() { }

        /// <summary>
        /// Initializes a new instance of the <see cref="Tree"/> class.
        /// </summary>
        /// <param name="id">The tree identifier.</param>
        /// <param name="species">The tree species.</param>
        public Tree(int id, TreeSpecies species)
        {
            Id = id;
            LatinName = species.LatinName;
            CommonName = species.CommonName;
        }

        /// <inheritdoc/>
        public int CompareTo(object? obj)
        {
            return CompareTo(obj as Tree);
        }

        /// <inheritdoc/>
        public int CompareTo(Tree? other)
        {
            if (other == null)
            {
                return 1;
            }
            else
            {
                int result = StringComparer.OrdinalIgnoreCase.Compare(CommonName, other.CommonName);

                if (result == 0)
                {
                    result = _id.CompareTo(other._id);
                }

                return result;
            }
        }

        /// <inheritdoc/>
        public override bool Equals(object? obj)
        {
            return Equals(obj as Tree);
        }

        /// <inheritdoc/>
        public bool Equals(Tree? other)
        {
            return other != null && _id == other._id && CommonName.Equals(other.CommonName, StringComparison.OrdinalIgnoreCase) && LatinName.Equals(other.LatinName, StringComparison.OrdinalIgnoreCase);
        }

        /// <inheritdoc/>
        public override int GetHashCode()
        {
            HashCode result = new HashCode();

            result.Add(_id);
            result.Add(CommonName);
            result.Add(LatinName);

            return result.ToHashCode();
        }

        /// <inheritdoc/>
        public override string ToString()
        {
            return ToString(formatProvider: null);
        }

        /// <inheritdoc/>
        public string ToString(IFormatProvider? formatProvider)
        {
            return ToString(format: null, formatProvider);
        }

        /// <inheritdoc/>
        public string ToString(string? format, IFormatProvider? formatProvider)
        {
            return string.Format(formatProvider, "{0} ({1}) #{2}", CommonName, LatinName, _id);
        }
    }
}
